File: android-compose/app/src/main/java/com/ohmz/tday/compose/core/data/server/ServerConfigRepository.kt
package com.ohmz.tday.compose.core.data.server

import android.util.Log
import com.ohmz.tday.compose.BuildConfig
import com.ohmz.tday.compose.core.data.AppDataMode
import com.ohmz.tday.compose.core.data.SecureConfigStore
import com.ohmz.tday.compose.core.data.ServerProbeException
import com.ohmz.tday.compose.core.data.extractApiErrorMessage
import com.ohmz.tday.compose.core.network.ServerTrustManager
import com.ohmz.tday.compose.core.network.isPrivateNetworkHost
import com.ohmz.tday.compose.core.network.TdayApiService
import com.ohmz.tday.compose.core.security.ProbeDecryptor
import kotlinx.coroutines.withTimeout
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import javax.inject.Inject
import javax.inject.Singleton
import javax.net.ssl.SSLException

@Singleton
class ServerConfigRepository @Inject constructor(
    private val api: TdayApiService,
    private val secureConfigStore: SecureConfigStore,
    private val serverTrustManager: ServerTrustManager,
) {
    fun getAppDataMode(): AppDataMode = secureConfigStore.getAppDataMode()

    fun isLocalMode(): Boolean = secureConfigStore.isLocalMode()

    fun hasServerConfigured(): Boolean = secureConfigStore.hasServerUrl()

    fun getServerUrl(): String? = secureConfigStore.getServerUrl()

    fun enableLocalMode() {
        secureConfigStore.clearServerUrl()
        secureConfigStore.clearCachedSessionUser()
        secureConfigStore.clearLastUsername()
        secureConfigStore.setAppDataMode(AppDataMode.LOCAL)
    }

    data class ProbeResult(
        val serverUrl: String,
        val versionCheck: VersionCheckResult,
        val backendVersion: String?,
    )

    suspend fun saveServerUrl(rawUrl: String): Result<String> =
        probeAndSave(rawUrl).map { it.serverUrl }

    suspend fun probeAndSave(rawUrl: String): Result<ProbeResult> = runCatching {
        val normalizedServerUrl = secureConfigStore.normalizeServerUrl(rawUrl)
            ?: throw ServerProbeException.InvalidUrl()
        val parsedServerUrl = normalizedServerUrl.toHttpUrlOrNull()
            ?: throw ServerProbeException.InvalidUrl()

        ensureSecureTransport(parsedServerUrl)

        val probeUrl = parsedServerUrl.newBuilder()
            .encodedPath(PROBE_PATH)
            .query(null)
            .fragment(null)
            .build()
            .toString()

        val probeResponse = try {
            withTimeout(PROBE_TIMEOUT_MS) {
                api.probeServer(probeUrl = probeUrl)
            }
        } catch (error: Exception) {
            throw describeTrustFailure(parsedServerUrl, error)
        }

        if (!probeResponse.isSuccessful) {
            throw IllegalStateException(
                extractApiErrorMessage(
                    probeResponse,
                    "Could not verify server. Check URL and try again.",
                ),
            )
        }

        val probeBody = probeResponse.body()
            ?: throw ServerProbeException.NotTdayServer()

        validateProbeContract(probeBody)

        val saved = secureConfigStore.saveServerUrl(
            rawUrl = normalizedServerUrl,
            persist = true,
        ).getOrThrow()
        secureConfigStore.clearOfflineSyncState()

        val compatibility = probeBody.encryptedCompatibility?.let { ProbeDecryptor.decrypt(it) }
        val versionCheck = checkVersionCompatibility(compatibility)

        ProbeResult(
            serverUrl = saved,
            versionCheck = versionCheck,
            backendVersion = compatibility?.appVersion ?: probeBody.appVersion,
        )
    }

    data class VersionRecheckResult(
        val versionCheck: VersionCheckResult,
        val serverAppVersion: String?,
    )

    suspend fun recheckVersion(): VersionRecheckResult {
        val serverUrl = getServerUrl()
            ?: return VersionRecheckResult(VersionCheckResult.Compatible, null)
        val parsedServerUrl = serverUrl.toHttpUrlOrNull()
            ?: return VersionRecheckResult(VersionCheckResult.Compatible, null)
        val probeUrl = parsedServerUrl.newBuilder()
            .encodedPath(PROBE_PATH)
            .query(null)
            .fragment(null)
            .build()
            .toString()

        val probeResponse = withTimeout(PROBE_TIMEOUT_MS) {
            api.probeServer(probeUrl = probeUrl)
        }

        val body = probeResponse.body()
            ?: throw IllegalStateException("Empty probe response during version recheck")
        val compatibility = body.encryptedCompatibility?.let { ProbeDecryptor.decrypt(it) }
        return VersionRecheckResult(
            versionCheck = checkVersionCompatibility(compatibility),
            serverAppVersion = compatibility?.appVersion ?: body.appVersion,
        )
    }

    fun resetTrustedServer(rawUrl: String): Result<Unit> {
        return secureConfigStore.clearTrustedServerFingerprintForUrl(rawUrl)
    }

    /**
     * Probes again with a one-shot authorisation to pin [fingerprint] — the value the user just
     * saw and confirmed. The authorisation names the expected certificate, so one swapped between
     * the prompt and this retry is still refused.
     */
    suspend fun confirmServerTrust(rawUrl: String, fingerprint: String): Result<ProbeResult> {
        val serverTrustKey = secureConfigStore.serverTrustKeyForUrl(rawUrl)
            ?: return Result.failure(ServerProbeException.InvalidUrl())

        serverTrustManager.allowEnrollment(serverTrustKey, fingerprint)
        val result = probeAndSave(rawUrl)
        if (result.isFailure) {
            serverTrustManager.cancelEnrollment(serverTrustKey)
        }
        return result
    }

    private fun validateProbeContract(probeBody: com.ohmz.tday.compose.core.model.MobileProbeResponse) {
        val serviceOk = probeBody.service.equals("tday", ignoreCase = true)
        val versionOk = probeBody.version == "1"

        if (serviceOk && versionOk) return

        Log.w(
            LOG_TAG,
            "probe_failed_contract service=${probeBody.service} version=${probeBody.version} probe=${probeBody.probe}",
        )
        throw ServerProbeException.NotTdayServer()
    }

    /**
     * Turns a refused handshake into the typed error the setup screen acts on: a changed
     * certificate (stored pin did not match) or an unrecognised one, carrying the fingerprint the
     * user has to confirm. Anything else — timeouts, DNS, plain connection failures — is rethrown
     * untouched.
     */
    private fun describeTrustFailure(serverUrl: HttpUrl, error: Exception): Exception {
        if (serverUrl.scheme != "https") return error
        if (!error.isTlsFailure()) return error

        val serverTrustKey = secureConfigStore.serverTrustKeyForUrl(serverUrl.toString())
            ?: return error

        if (serverTrustManager.consumeMismatch(serverTrustKey)) {
            return ServerProbeException.CertificateChanged(serverTrustKey)
        }

        val offeredFingerprint = serverTrustManager.consumeUnknownFingerprint(serverTrustKey)
            ?: return error

        return ServerProbeException.CertificateUntrusted(serverTrustKey, offeredFingerprint)
    }

    /// Handshake refusals surface as an SSLException, sometimes wrapped by OkHttp's route retry.
    private fun Throwable.isTlsFailure(): Boolean {
        var current: Throwable? = this
        while (current != null) {
            if (current is SSLException) return true
            current = current.cause?.takeIf { it !== current }
        }
        return false
    }

    private fun ensureSecureTransport(serverUrl: HttpUrl) {
        if (serverUrl.scheme == "https") return
        // Shares isPrivateNetworkHost with the trust manager on purpose: "is this host reachable
        // only from the LAN?" now gates both cleartext and certificate enrollment, and a second
        // copy of the rule is how the two drift apart.
        if (BuildConfig.DEBUG && isPrivateNetworkHost(serverUrl.host)) return
        throw ServerProbeException.InsecureTransport()
    }

    private companion object {
        const val LOG_TAG = "ServerConfigRepo"
        const val PROBE_TIMEOUT_MS = 7_000L
        const val PROBE_PATH = "/api/mobile/probe"
    }
}

File: ios-swiftUI/Tday/Core/Data/Server/ServerConfigRepository.swift
import Foundation

enum ServerProbeError: Error, Equatable, LocalizedError {
    case invalidURL
    case insecureTransport
    case notTdayServer
    case certificateChanged
    /// The server presented a certificate the system does not trust and that we have never pinned.
    /// Carries the fingerprint so the UI can show it and ask the user to confirm. Only ever raised
    /// for private/LAN hosts — see `untrustedPublicCertificate`.
    case untrustedCertificate(host: String, fingerprint: String?)
    /// A PUBLIC host presented a certificate the system cannot verify. Carries no fingerprint on
    /// purpose: there is nothing for the user to approve, because approving it is the attack.
    case untrustedPublicCertificate(host: String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Enter a valid server URL"
        case .insecureTransport:
            return "HTTPS is required for remote servers"
        case .notTdayServer:
            return "This server does not look like a T'Day instance"
        case .certificateChanged:
            return "The trusted certificate changed for this server"
        case .untrustedCertificate:
            return "This server's certificate is not trusted"
        case .untrustedPublicCertificate:
            return "This server's certificate could not be verified"
        }
    }
}

final class ServerConfigRepository {
    private let secureStore: SecureStore
    private let serverURLState: ServerURLState
    private let api: TdayAPIService

    init(secureStore: SecureStore, serverURLState: ServerURLState, api: TdayAPIService) {
        self.secureStore = secureStore
        self.serverURLState = serverURLState
        self.api = api
    }

    func hasServerConfigured() -> Bool {
        serverURLState.currentURL != nil || secureStore.loadPersistedServerURL() != nil
    }

    func appDataMode() -> AppDataMode {
        secureStore.appDataMode()
    }

    func isLocalMode() -> Bool {
        secureStore.isLocalMode()
    }

    func getServerURL() -> URL? {
        serverURLState.currentURL ?? secureStore.loadPersistedServerURL()
    }

    func serverURL() -> String? {
        getServerURL()?.absoluteString
    }

    struct ProbeResult {
        let serverURL: String
        let versionCheck: VersionCheckResult
        let backendVersion: String?
    }

    struct VersionRecheckResult {
        let versionCheck: VersionCheckResult
        let backendVersion: String?
        /// True when the probe was refused by the TLS trust check rather than merely failing.
        /// Without this the caller cannot tell "server unreachable" from "someone may be
        /// intercepting this connection", and both would report `.compatible`.
        var trustFailed: Bool = false
    }

    func saveServerURL(rawURL: String) async throws -> MobileProbeResponse {
        let result = try await probe(rawURL: rawURL)
        let response = result.response
        guard response.service.compare("tday", options: .caseInsensitive) == .orderedSame,
              response.version == "1" else {
            throw ServerProbeError.notTdayServer
        }
        serverURLState.currentURL = result.serverURL
        persistRuntimeServerURL()
        return response
    }

    func saveServerURL(_ rawURL: String) async throws -> String {
        _ = try await saveServerURL(rawURL: rawURL)
        return getServerURL()?.absoluteString ?? rawURL
    }

    func probeAndSave(_ rawURL: String) async throws -> ProbeResult {
        let result = try await probe(rawURL: rawURL)
        let response = result.response
        guard response.service.compare("tday", options: .caseInsensitive) == .orderedSame,
              response.version == "1" else {
            throw ServerProbeError.notTdayServer
        }
        serverURLState.currentURL = result.serverURL
        persistRuntimeServerURL()
        let compatibility = response.encryptedCompatibility.flatMap { ProbeDecryptor.decrypt($0) }
        let versionCheck = checkVersionCompatibility(payload: compatibility)
        return ProbeResult(
            serverURL: result.serverURL.absoluteString,
            versionCheck: versionCheck,
            backendVersion: compatibility?.appVersion ?? response.appVersion
        )
    }

    func recheckVersion() async -> VersionRecheckResult {
        guard let url = getServerURL() else {
            return VersionRecheckResult(versionCheck: .compatible, backendVersion: nil)
        }
        let probeURL = url.appending(path: "api/mobile/probe")
        guard let response = try? await api.probeServer(url: probeURL) else {
            // Now that the trust check fails closed, a refused certificate reaches this path as a
            // bare cancellation. Reporting ".compatible" here would silently mask exactly the
            // interception the fail-closed change exists to catch.
            let trustFailed = url.host.map { host in
                api.consumeTrustFailure(forHost: host) ||
                    api.consumePublicTrustRefusal(forHost: host) ||
                    api.consumeTrustUnknown(forHost: host) != nil
            } ?? false
            return VersionRecheckResult(versionCheck: .compatible, backendVersion: nil, trustFailed: trustFailed)
        }
        let compatibility = response.encryptedCompatibility.flatMap { ProbeDecryptor.decrypt($0) }
        return VersionRecheckResult(
            versionCheck: checkVersionCompatibility(payload: compatibility),
            backendVersion: compatibility?.appVersion ?? response.appVersion
        )
    }

    /// Forgets the stored pin and re-probes.
    ///
    /// Dropping the pin does not mean "trust whatever answers next": the probe goes through the
    /// same fail-closed path, so an unrecognised certificate comes back as
    /// [ServerProbeError.untrustedCertificate] and the user still has to confirm the fingerprint.
    func resetTrustedServer(rawURL: String) async throws -> MobileProbeResponse {
        let normalizedURL = try normalize(rawURL: rawURL)
        if let host = normalizedURL.host {
            secureStore.clearTrustedFingerprint(for: host)
        }
        let (serverURL, response) = try await probe(rawURL: rawURL)
        guard response.service.compare("tday", options: .caseInsensitive) == .orderedSame,
              response.version == "1" else {
            throw ServerProbeError.notTdayServer
        }
        serverURLState.currentURL = serverURL
        persistRuntimeServerURL()
        return response
    }

    func persistRuntimeServerURL() {
        guard let url = serverURLState.currentURL else {
            return
        }
        secureStore.savePersistedServerURL(url)
    }

    func clearServerConfiguration() {
        serverURLState.currentURL = nil
        secureStore.clearPersistedServerURL()
        secureStore.clearAppDataMode()
        secureStore.clearAllTrustedFingerprints()
    }

    func enableLocalMode() {
        serverURLState.currentURL = nil
        secureStore.clearPersistedServerURL()
        secureStore.clearCachedSessionUser()
        secureStore.clearLastUsername()
        secureStore.clearPersistedAuthSessionCookie()
        secureStore.setAppDataMode(.local)
    }

    /// The exact inverse of [enableLocalMode], and deliberately nothing more.
    ///
    /// Leaving a local workspace is a MODE SWITCH, not a teardown — web says so in as many
    /// words (`AuthProvider.tsx`: "Leaving the local workspace is a mode switch, not a session
    /// teardown"). The rows stay on the device so choosing Local Mode again finds the
    /// workspace where it was left. Wiping is the separate, confirmed row next to it; the
    /// marker is what stops `bootstrap()` folding the two together on the next launch.
    func leaveLocalMode() {
        secureStore.clearCachedSessionUser()
        secureStore.setAppDataMode(.unset)
        secureStore.setRetainedLocalWorkspace(true)
    }

    /// A local workspace that was left rather than deleted is still on disk.
    var hasRetainedLocalWorkspace: Bool {
        secureStore.hasRetainedLocalWorkspace()
    }

    func clearRetainedLocalWorkspace() {
        secureStore.setRetainedLocalWorkspace(false)
    }

    func buildAbsoluteAppURL(_ path: String) -> URL? {
        getServerURL()?.appending(path: path.trimmingCharacters(in: CharacterSet(charactersIn: "/")))
    }

    private func normalize(rawURL: String) throws -> URL {
        let trimmed = rawURL.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw APIError(message: "Server URL is required", statusCode: nil)
        }

        let candidate = trimmed.contains("://") ? trimmed : "https://\(trimmed)"
        guard var components = URLComponents(string: candidate), let host = components.host?.lowercased() else {
            throw ServerProbeError.invalidURL
        }

        let scheme = components.scheme?.lowercased() ?? "https"
        guard scheme == "https" || scheme == "http" else {
            throw ServerProbeError.invalidURL
        }
        if scheme == "http" && !isLocalAddress(host) {
            throw ServerProbeError.insecureTransport
        }

        components.scheme = scheme
        components.path = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.query = nil
        components.fragment = nil

        guard let resolvedURL = components.url else {
            throw ServerProbeError.invalidURL
        }
        return resolvedURL
    }

    private func probe(rawURL: String) async throws -> (serverURL: URL, response: MobileProbeResponse) {
        let normalizedURL = try normalize(rawURL: rawURL)
        do {
            let response = try await api.probeServer(url: normalizedURL.appending(path: "api/mobile/probe"))
            return (normalizedURL, response)
        } catch {
            // The TLS pinning check cancels the request (URLError.cancelled) when the
            // server's certificate no longer matches the pinned fingerprint. Translate
            // that into a clear, actionable error instead of a bare "cancelled".
            if let host = normalizedURL.host, api.consumeTrustFailure(forHost: host) {
                throw ServerProbeError.certificateChanged
            }
            // Public host we could not verify: refused with no fingerprint to approve. Checked
            // before the enrollment path below so this can never fall through into a trust prompt.
            if let host = normalizedURL.host, api.consumePublicTrustRefusal(forHost: host) {
                throw ServerProbeError.untrustedPublicCertificate(host: host)
            }
            // Unrecognised certificate on a private/LAN host: refused rather than
            // trusted-on-first-use. Surface the fingerprint so the setup screen can ask the user to
            // confirm it explicitly. The outer `if let` only tests that a refusal was recorded —
            // the fingerprint itself may be nil.
            if let host = normalizedURL.host, let fingerprint = api.consumeTrustUnknown(forHost: host) {
                throw ServerProbeError.untrustedCertificate(host: host, fingerprint: fingerprint)
            }
            throw error
        }
    }

    /// Pins `fingerprint` for `host` and re-probes, after the user has confirmed it on screen.
    ///
    /// The approval names the exact fingerprint, so a certificate swapped between the prompt and
    /// this retry is still refused — this is a confirmation, not a blanket "trust anything next".
    func approveCertificate(rawURL: String, host: String, fingerprint: String) async throws -> MobileProbeResponse {
        api.allowTrustEnrollment(host: host, expecting: fingerprint)
        do {
            let (serverURL, response) = try await probe(rawURL: rawURL)
            guard response.service.compare("tday", options: .caseInsensitive) == .orderedSame,
                  response.version == "1" else {
                throw ServerProbeError.notTdayServer
            }
            serverURLState.currentURL = serverURL
            persistRuntimeServerURL()
            return response
        } catch {
            api.cancelTrustEnrollment(host: host)
            throw error
        }
    }

    private func isLocalAddress(_ host: String) -> Bool {
        host == "localhost" ||
        host == "127.0.0.1" ||
        host == "10.0.2.2" ||
        host.hasPrefix("192.168.") ||
        host.hasPrefix("10.") ||
        host.hasSuffix(".local")
    }
}
